File: Problems/01984. Minimum Difference Between Highest and Lowest of K Scores/C++/Solution.cpp
class Solution {
    public:
        int minimumDifference(vector<int>& nums, int k) {
            cout << "Original nums: ";
            for (int n : nums) {
                cout << n << " ";
            }
            cout << "\nValue of k: " << k << endl << endl;
    
            // Edge Case
            if (k == 1) {
                cout << "Since k = 1, the min difference is 0." << endl;
                return 0;
            }
    
            // Sort Array
            mergeSort(nums, 0, nums.size() - 1);
    
            cout << "Sorted: ";
            for (int n : nums) {
                cout << n << " ";
            }
            cout << endl << endl;
    
            // Initialize minDiff
            int minDiff = INT_MAX;
    
            // Use Sliding Window
            for (int i = 0; i <= nums.size() - k; ++i) {
                int windowStart = nums[i];
                int windowEnd = nums[i + k - 1];
                int diff = windowEnd - windowStart;
    
                cout << "Window [" << i << " to " << i + k - 1 << "] -> (" << windowEnd << " - " << windowStart << ") = " << diff << endl;
    
                if (diff < minDiff) {
                    cout << "--> New minDiff found: " << diff << endl << endl;
                    minDiff = diff;
                }
            }
    
            // Return result
            cout << endl << "Final minimum difference: " << minDiff << endl;
            return minDiff;
    
        }
    
        void mergeSort(vector<int>& arr, int left, int right) {
            if (left >= right) {
                return;
            }
    
            int mid = left + (right - left) / 2;
    
            mergeSort(arr, left, mid);
            mergeSort(arr, mid + 1, right);
    
            merge(arr, left, mid, right);
        }
    
        void merge(vector<int>& arr, int left, int mid, int right) {
            int n1 = mid - left + 1;
            int n2 = right - mid;
    
            vector<int> L(n1);
            vector<int> R(n2);
    
            for (int i = 0; i < n1; ++i) {
                L[i] = arr[left + i];
            }
            for (int j = 0; j < n2; ++j) {
                R[j] = arr[mid + 1 + j];
            }
    
            int i = 0;
            int j = 0;
            int k = left;
    
            while(i < n1 && j < n2) {
                if(L[i] <= R[j]) {
                    arr[k++] = L[i++];
                } else {
                    arr[k++] = R[j++];
                }
            }
    
            while (i < n1) {
                arr[k++] = L[i++];
            }
            while (j < n2) {
                arr[k++] = R[j++];
            }
        }
    
    };

File: Problems/01984. Minimum Difference Between Highest and Lowest of K Scores/TypeScript/Solution.ts
function minimumDifference(nums: number[], k: number): number {
    console.log("Original nums:", nums);
    console.log("Value of k:", k);

    // Edge Case
    if (k === 1) {
        console.log("Since k = 1, the minimum difference is 0.");
        return 0;
    }

    // Sort Array
    nums = mergeSort(nums);
    console.log("Sorted:", nums);

    // Initialize minDiff
    let minDiff = Infinity;

    // Use Sliding Window
    for (let i = 0; i < nums.length - k + 1; i++) {
        const windowStart = nums[i];
        const windowEnd = nums[i + k - 1];
        const currentDiff = windowEnd - windowStart;

        console.log(`Window [${i} to ${i + k - 1}] -> (${windowEnd} - ${windowStart}) = ${currentDiff}`);

        if (currentDiff < minDiff) {
            console.log(`--> New minDiff found: ${currentDiff}`);
            minDiff = currentDiff;
        }
    }

    // Return result
    console.log("Final minimum difference:", minDiff);
    return minDiff;
};

function mergeSort(arr: number[]): number[] {
    if (arr.length < 2) {
        return arr;
    }

    const mid = Math.floor(arr.length / 2);
    const left = mergeSort(arr.slice(0, mid));
    const right = mergeSort(arr.slice(mid));

    return merge(left, right);
}

function merge(left: number[], right: number[]): number[] {
    let result: number[] = [];
    let i = 0;
    let j = 0;

    while (i < left.length && j < right.length) {
        if (left[i] <= right[j]) {
            result.push(left[i++]);
        } else {
            result.push(right[j++]);
        }
    }
    return result.concat(left.slice(i)).concat(right.slice(j));
}